File: custom-plugins/cordova-plugin-wifi-billing-agent/src/ios/agents/AuthAgent.m
//
//  AuthAgent.m
//  WiFi Billing Agent
//

#import "AuthAgent.h"
#import "CoreAgent.h"

@interface AuthAgent()

@property (nonatomic, strong) NSString *apiEndpoint;
@property (nonatomic, strong) NSString *apiKey;
@property (nonatomic, strong) NSString *currentToken;
@property (nonatomic, strong) NSString *currentUserId;
@property (nonatomic, assign) NSTimeInterval tokenExpiry;
@property (nonatomic, strong) NSTimer *refreshTimer;

@end

@implementation AuthAgent

- (instancetype)initWithConfig:(NSDictionary *)config {
    self = [super init];
    if (self) {
        _apiEndpoint = config[@"apiEndpoint"] ?: @"";
        _apiKey = config[@"apiKey"] ?: @"";
        
        // Load from keychain if available
        [self loadStoredCredentials];
    }
    return self;
}

- (void)initialize {
    NSLog(@"[AuthAgent] Initializing");
    
    // Check if we have a valid token
    if (self.currentToken && self.tokenExpiry > [[NSDate date] timeIntervalSince1970]) {
        NSLog(@"[AuthAgent] Found valid token");
        [self scheduleTokenRefresh];
    }
}

- (void)shutdown {
    NSLog(@"[AuthAgent] Shutting down");
    [self.refreshTimer invalidate];
}

- (NSDictionary *)authenticate:(NSDictionary *)credentials {
    NSString *username = credentials[@"username"];
    NSString *password = credentials[@"password"];
    NSString *deviceId = credentials[@"deviceId"];
    
    // In production, call your backend API
    // For now, simulate successful auth
    
    self.currentToken = [self generateToken:username];
    self.currentUserId = username;
    self.tokenExpiry = [[NSDate date] timeIntervalSince1970] + (24 * 60 * 60); // 24 hours
    
    [self saveCredentials];
    [self scheduleTokenRefresh];
    
    NSLog(@"[AuthAgent] Authenticated user: %@", username);
    
    return @{
        @"success": @(YES),
        @"token": self.currentToken,
        @"userId": username,
        @"expiry": @(self.tokenExpiry),
        @"deviceProfiled": @(YES)
    };
}

- (NSDictionary *)validateToken:(NSString *)token {
    if (!self.currentToken || ![self.currentToken isEqualToString:token]) {
        return @{
            @"valid": @(NO),
            @"reason": @"invalid_token"
        };
    }
    
    if ([[NSDate date] timeIntervalSince1970] > self.tokenExpiry) {
        return @{
            @"valid": @(NO),
            @"reason": @"expired"
        };
    }
    
    return @{
        @"valid": @(YES),
        @"userId": self.currentUserId,
        @"expiresIn": @(self.tokenExpiry - [[NSDate date] timeIntervalSince1970])
    };
}

- (NSDictionary *)refreshToken {
    if (!self.currentToken || !self.currentUserId) {
        return @{
            @"success": @(NO),
            @"error": @"No active session"
        };
    }
    
    // Generate new token
    self.currentToken = [self generateToken:self.currentUserId];
    self.tokenExpiry = [[NSDate date] timeIntervalSince1970] + (24 * 60 * 60);
    
    [self saveCredentials];
    [self scheduleTokenRefresh];
    
    NSLog(@"[AuthAgent] Token refreshed");
    
    return @{
        @"success": @(YES),
        @"token": self.currentToken,
        @"expiry": @(self.tokenExpiry)
    };
}

- (void)logout {
    self.currentToken = nil;
    self.currentUserId = nil;
    self.tokenExpiry = 0;
    
    [self.refreshTimer invalidate];
    [self clearStoredCredentials];
    
    NSLog(@"[AuthAgent] Logged out");
}

- (NSString *)generateToken:(NSString *)username {
    NSTimeInterval timestamp = [[NSDate date] timeIntervalSince1970] * 1000;
    NSString *random = [[NSUUID UUID] UUIDString];
    NSString *combined = [NSString stringWithFormat:@"pc-tok-%.0f-%@-%@", timestamp, username, random];
    
    // Base64 encode
    NSData *data = [combined dataUsingEncoding:NSUTF8StringEncoding];
    return [data base64EncodedStringWithOptions:0];
}

- (void)scheduleTokenRefresh {
    [self.refreshTimer invalidate];
    
    // Refresh 5 minutes before expiry
    NSTimeInterval refreshInterval = self.tokenExpiry - [[NSDate date] timeIntervalSince1970] - (5 * 60);
    
    if (refreshInterval > 0) {
        self.refreshTimer = [NSTimer scheduledTimerWithTimeInterval:refreshInterval
                                                             target:self
                                                           selector:@selector(refreshToken)
                                                           userInfo:nil
                                                            repeats:NO];
    }
}

- (void)saveCredentials {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    [defaults setObject:self.currentToken forKey:@"wba_auth_token"];
    [defaults setObject:self.currentUserId forKey:@"wba_user_id"];
    [defaults setDouble:self.tokenExpiry forKey:@"wba_token_expiry"];
    [defaults synchronize];
}

- (void)loadStoredCredentials {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    self.currentToken = [defaults objectForKey:@"wba_auth_token"];
    self.currentUserId = [defaults objectForKey:@"wba_user_id"];
    self.tokenExpiry = [defaults doubleForKey:@"wba_token_expiry"];
}

- (void)clearStoredCredentials {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    [defaults removeObjectForKey:@"wba_auth_token"];
    [defaults removeObjectForKey:@"wba_user_id"];
    [defaults removeObjectForKey:@"wba_token_expiry"];
    [defaults synchronize];
}

@end

File: custom-plugins/cordova-plugin-aicore/cordova-plugin-wifi-billing-agent/src/android/agents/AuthAgent.java
package com.wifibilling.agent.agents;

import android.content.Context;
import android.content.SharedPreferences;

import org.json.JSONException;
import org.json.JSONObject;

public class AuthAgent {

    private static final String PREFS_NAME = "WiFiBillingAuth";
    private static final String KEY_TOKEN = "auth_token";
    private static final String KEY_USER_ID = "user_id";
    private static final String KEY_EXPIRY = "token_expiry";

    private Context context;
    private CoreAgent coreAgent;
    private SharedPreferences prefs;
    private String apiEndpoint;
    private String apiKey;

    public AuthAgent(Context context, CoreAgent coreAgent, JSONObject config) {
        this.context = context;
        this.coreAgent = coreAgent;
        this.prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        this.apiEndpoint = config.optString("apiEndpoint", "");
        this.apiKey = config.optString("apiKey", "");
    }

    public void initialize() {
        // Check for existing valid token
    }

    public JSONObject authenticate(JSONObject credentials) throws JSONException {
        String username = credentials.optString("username");
        String password = credentials.optString("password");
        String deviceId = credentials.optString("deviceId");

        // In production, this would call your backend API
        // For now, simulate successful authentication

        String token = generateToken(username);
        long expiry = System.currentTimeMillis() + (24 * 60 * 60 * 1000); // 24 hours

        SharedPreferences.Editor editor = prefs.edit();
        editor.putString(KEY_TOKEN, token);
        editor.putString(KEY_USER_ID, username);
        editor.putLong(KEY_EXPIRY, expiry);
        editor.apply();

        JSONObject result = new JSONObject();
        result.put("success", true);
        result.put("token", token);
        result.put("userId", username);
        result.put("expiry", expiry);
        result.put("deviceProfiled", true);

        return result;
    }

    public JSONObject validateToken(String token) throws JSONException {
        String storedToken = prefs.getString(KEY_TOKEN, null);
        long expiry = prefs.getLong(KEY_EXPIRY, 0);

        JSONObject result = new JSONObject();

        if (storedToken == null || !storedToken.equals(token)) {
            result.put("valid", false);
            result.put("reason", "invalid_token");
        } else if (System.currentTimeMillis() > expiry) {
            result.put("valid", false);
            result.put("reason", "expired");
        } else {
            result.put("valid", true);
            result.put("userId", prefs.getString(KEY_USER_ID, null));
            result.put("expiresIn", (expiry - System.currentTimeMillis()) / 1000);
        }

        return result;
    }

    public JSONObject refreshToken() throws JSONException {
        String currentToken = prefs.getString(KEY_TOKEN, null);
        String userId = prefs.getString(KEY_USER_ID, null);

        if (currentToken == null || userId == null) {
            throw new JSONException("No active session to refresh");
        }

        // Generate new token
        String newToken = generateToken(userId);
        long newExpiry = System.currentTimeMillis() + (24 * 60 * 60 * 1000);

        SharedPreferences.Editor editor = prefs.edit();
        editor.putString(KEY_TOKEN, newToken);
        editor.putLong(KEY_EXPIRY, newExpiry);
        editor.apply();

        JSONObject result = new JSONObject();
        result.put("success", true);
        result.put("token", newToken);
        result.put("expiry", newExpiry);

        return result;
    }

    public void logout() {
        SharedPreferences.Editor editor = prefs.edit();
        editor.clear();
        editor.apply();
    }

    private String generateToken(String username) {
        return "pc-tok-" + System.currentTimeMillis() + "-" +
                android.util.Base64.encodeToString(username.getBytes(), android.util.Base64.URL_SAFE);
    }

    public void shutdown() {
        // Cleanup
    }
}